x/staking/handler.go

```diff
@@ -245,8 +255,19 @@ func handleMsgUndelegate(ctx sdk.Context, msg types.MsgUndelegate, k keeper.Keep
 }
 
 func handleMsgBeginRedelegate(ctx sdk.Context, msg types.MsgBeginRedelegate, k keeper.Keeper) sdk.Result {
-	completionTime, err := k.BeginRedelegation(ctx, msg.DelegatorAddress, msg.ValidatorSrcAddress,
-		msg.ValidatorDstAddress, msg.SharesAmount)
+	validator, found := k.GetValidator(ctx, msg.ValidatorSrcAddress)
+	if !found {
+		return ErrNoValidatorFound(k.Codespace()).Result()
+	}
+
+	if validator.Tokens.IsZero() {
+		return ErrInsufficientShares(k.Codespace()).Result()
+	}
+
+	sharesAmt := validator.DelegatorShares.MulInt(msg.Amount.Amount).QuoInt(validator.Tokens).TruncateDec()
```

Comment:
ditto on _all_ three previous points from handle `handleMsgUndelegate` - all the logic is the same, why not consolidate into a child function like `GetSharesAmount(...) error`

Reply:
ACKed.